src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -141,10 +142,24 @@ public function dump(array $options = array())
 
         (new AnalyzeServiceReferencesPass())->process($this->container);
         $this->circularReferences = array();
+        $this->singlePrivateIds = array();
         $checkedNodes = array();
         foreach ($this->container->getCompiler()->getServiceReferenceGraph()->getNodes() as $id => $node) {
             $currentPath = array($id => $id);
             $this->analyzeCircularReferences($node->getOutEdges(), $checkedNodes, $currentPath);
+            if ($node->getValue()->isPublic()) {
+                continue;
+            }
+            $ids = array();
+            foreach ($node->getInEdges() as $edge) {
```

Comment:
I'd recommend to extract that whole logic into a new private method for clarity.

Reply:
done